pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	selectors, err := api.SelectorsFromProto(req.Selectors)
+	if err != nil {
+		return err
+	}
+
+	subscriber := s.manager.SubscribeToCacheChanges(selectors)
+	defer subscriber.Finish()
+
+	for {
+		select {
+		case update := <-subscriber.Updates():
+			if err := sendX509SVIDResponse(update, stream, log); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sendX509SVIDResponse(update *cache.WorkloadUpdate, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer, log logrus.FieldLogger) (err error) {
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Error(codes.Unavailable, "could not serialize response")
+	}
+
+	if err := stream.Send(resp); err != nil {
+		log.WithError(err).Error("Failed to send X.509 SVID response")
+		return err
+	}
+
+	return nil
+}
+
+func composeX509SVIDBySelectors(update *cache.WorkloadUpdate) (*delegatedidentityv1.SubscribeToX509SVIDsResponse, error) {
+	resp := new(delegatedidentityv1.SubscribeToX509SVIDsResponse)
+	resp.X509Svids = []*delegatedidentityv1.X509SVIDWithKey{}
+
+	for td, _ := range update.FederatedBundles {
+		resp.FederatesWith = append(resp.FederatesWith, td.IDString())
+	}
+
+	for _, identity := range update.Identities {
+		// TODO:  it doesn't work and always prints "false"
+		//fmt.Printf("processing id %s, %t\n", identity.Entry.SpiffeId, identity.Entry.Admin)
+		// Do not send admin nor downstream SVIDs to the caller
+		if identity.Entry.Admin || identity.Entry.Downstream {
+			continue
+		}
+
+		id, _ := idutil.IDProtoFromString(identity.Entry.SpiffeId)
+
+		keyData, err := x509.MarshalPKCS8PrivateKey(identity.PrivateKey)
+		if err != nil {
+			return nil, fmt.Errorf("marshal key for %v: %w", id, err)
+		}
+
+		svid := &delegatedidentityv1.X509SVIDWithKey{
+			X509Svid: &types.X509SVID{
+				Id:        id,
+				CertChain: x509util.RawCertsFromCertificates(identity.SVID),
+				// TODO: what if SVIDs 0 doesn't exist?
+				ExpiresAt: identity.SVID[0].NotAfter.Unix(),
```

Comment:
Based on this TODO, is there any use case where try to access index 0 could result in a problem?
I tried to find for some places where the behavior is the same and at workloadAPI `FetchX509SVID` I found the same behavior [here](https://github.com/spiffe/spire/blob/main/pkg/agent/endpoints/workload/handler.go#L336) and in this case no extra validation was applied when trying to access SVID[0].

Reply:
Similar to my above comment, I think it is very unlikely but probably worth checking.